### ISSUE TITLE
Archives are wrongly detected as duplicates

### ISSUE THREAD

**Issue by CirnoT** [Difegue/LANraragi#164]
**LRR Version and OS**
Current `dev` branch as of time of writing this (74751bdbcf39c3166da8fb2b5e3fe61050dca590).

**Bug Details**
Uploading archives via web uploader causes some of them to be wrongly detected as duplicates. This can be observed by checking Shinobu log. At one occasion I also found that archive recently uploaded by web uploader had thumbnail from some random archive added quite some time ago - regenerating fixed it. Additionally, removing the file and then copying it back fixed it but did not generate message 'Adding new file' in Shinobu logs. As for plugin logs, I can see
```
[2019-08-05 19:32:59] [E-Hentai] [info] Reverse Image Search Enabled, trying first.
[2019-08-05 19:32:59] [E-Hentai] [info] Sending the following tags to LRR: language:english, language:translated, parody:bakemonogatari, character:koyomi araragi, character:shinobu oshino, group:digital lover, artist:nakajima yuka, male:sole male, female:anal, female:lolicon, female:sole female, female:sundress, category:doujinshi
```
indicating that plugin execution was successful and correct on the archive. it looks like either filemap or JSON cache got corrupted.

**Matching Logs**
```
[2019-08-05 13:46:26] [Shinobu] [warn] /db/(C88) [French letter (Fujisaki Hikari)] 401 -Ano hi, Natsu no Owari ni- (Kantai Collection -KanColle-) [English] [arkeros].zip is a duplicate of the existing file /db/[Yuushi Tessen] Tadashii Koude no Tsukaikata (Heartcatch Puni Pedo! Cosmos Gumi) [English] {Hennojin}.zip. You should delete it.
[2019-08-05 19:33:05] [Shinobu] [warn] /db/(CSP5) [Asaiumi (Asami Asami)] Nadeko no Hon (Bakemonogatari) [English] =TV=.zip is a duplicate of the existing file /db/(SC54) [TIES (Takei Ooki)] Karen na Tsukihi (Nisemonogatari) [English] [DirtyTranslations].zip. You should delete it.
[2019-08-05 19:33:02] [Shinobu] [warn] /db/(COMIC1â7) [Galley (ryoma)] Sukimonogatari (Bakemonogatari) [English] =YQII=.zip is a duplicate of the existing file /db/(SC54) [TIES (Takei Ooki)] Karen na Tsukihi (Nisemonogatari) [English] [DirtyTranslations].zip. You should delete it.
[2019-08-05 19:33:00] [Shinobu] [warn] /db/(C82) [TIES (Takei Ooki)] Oniichan wa Imouto ni Yokujou Shitari Shinaiyone  (Bakemonogatari) [English] [life4Kaoru].zip is a duplicate of the existing file /db/(SC54) [TIES (Takei Ooki)] Karen na Tsukihi (Nisemonogatari) [English] [DirtyTranslations].zip. You should delete it.
[2019-08-05 19:32:58] [Shinobu] [warn] /db/(C84) [Digital Lover (Nakajima Yuka)] D.L. action 79 (Bakemonogatari) [English] [YQII].zip is a duplicate of the existing file /db/(SC54) [TIES (Takei Ooki)] Karen na Tsukihi (Nisemonogatari) [English] [DirtyTranslations].zip. You should delete it.
```

**Screenshots**
n/a

**Comment by CirnoT** [comment]
This is either web upload issue (I've started using it today) or issue caused by recent commit to disable rebuilding of filemap in Shinobu.

**Comment by Difegue** [comment]
Can you try with the pre-filemap disabling commit to see if that fixes the issue?  

The web uploader uses the same code as the content folder detection for the archive processing part, so I don't think the issue can come from that in particular.

**Comment by CirnoT** [comment]
I will duplicate the LXC container I'm using for LRR later and try playing with it there. For now I can say that since posting this issue I've uploaded files via SFTP normally without any issue. Seeing how they all were detected as duplicates of same archive it's very likely there's some issue with how the files are uploaded. Nevertheless, the files were uploaded properly and plugins ran on them as well as they were properly saved in database with their metadata from plugins, so it seems like duplication detection was the only thing preventing them from being shown.

**Comment by CirnoT** [comment]
Well, this is pretty weird

```
[2019-08-06 01:35:10] [LANraragi] [debug] Shinobu Worker new PID is 10322
[2019-08-06 01:35:10] [LANraragi] [debug] Routing done! Ready to receive requests.
[2019-08-06 01:35:51] [LANraragi] [debug] ID of uploaded file is 90c8c7a827cffbc59f5b7f4c6b45f51de93b2f16
[2019-08-06 01:35:51] [Archive] [debug] Pushing to redis on ID 90c8c7a827cffbc59f5b7f4c6b45f51de93b2f16:
[2019-08-06 01:35:51] [Archive] [debug] File Name: (C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo]
[2019-08-06 01:35:51] [Archive] [debug] Filesystem Path: /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip
[2019-08-06 01:35:51] [Auto-Plugin] [info] Executing enabled plugins on archive with id 90c8c7a827cffbc59f5b7f4c6b45f51de93b2f16.
[2019-08-06 01:35:51] [Auto-Tagger] [info] Thumbnail hash invalid, regenerating.
[2019-08-06 01:35:53] [LANraragi] [debug] ID of uploaded file is e8ca20b95b1a22af7661f866b34aca331284ddda
[2019-08-06 01:35:53] [Archive] [debug] Pushing to redis on ID e8ca20b95b1a22af7661f866b34aca331284ddda:
[2019-08-06 01:35:53] [Archive] [debug] File Name: [Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital]
[2019-08-06 01:35:53] [Archive] [debug] Filesystem Path: /db/[Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital].zip
[2019-08-06 01:35:53] [Auto-Plugin] [info] Executing enabled plugins on archive with id e8ca20b95b1a22af7661f866b34aca331284ddda.
[2019-08-06 01:35:53] [Auto-Tagger] [info] Thumbnail hash invalid, regenerating.

[2019-08-06 01:35:44] [Shinobu] [info] All done! Now dutifully watching your files. 
[2019-08-06 01:35:51] [Shinobu] [debug] Received inotify event 256 on /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip
[2019-08-06 01:35:51] [Shinobu] [debug] Adding /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip to Shinobu filemap.
[2019-08-06 01:35:51] [Shinobu] [debug] Computed ID is da39a3ee5e6b4b0d3255bfef95601890afd80709.
[2019-08-06 01:35:51] [Shinobu] [info] Building JSON cache...This might take some time.
[2019-08-06 01:35:51] [Shinobu] [debug] File name discrepancy detected between DB and filesystem!
[2019-08-06 01:35:51] [Shinobu] [debug] Filesystem: /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip
[2019-08-06 01:35:51] [Shinobu] [debug] Database: /db/(SC54) [TIES (Takei Ooki)] Karen na Tsukihi (Nisemonogatari) [English] [DirtyTranslations].zip
[2019-08-06 01:35:53] [Shinobu] [debug] Received inotify event 256 on /db/[Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital].zip
[2019-08-06 01:35:53] [Shinobu] [debug] Adding /db/[Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital].zip to Shinobu filemap.
[2019-08-06 01:35:53] [Shinobu] [debug] Computed ID is da39a3ee5e6b4b0d3255bfef95601890afd80709.
[2019-08-06 01:35:53] [Shinobu] [warn] /db/[Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital].zip is a duplicate of the existing file /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip. You should delete it.
[2019-08-06 01:35:53] [Shinobu] [info] Building JSON cache...This might take some time.
```

**Comment by CirnoT** [comment]
This should be clearer - after manually removing entry for da39a3ee5e6b4b0d3255bfef95601890afd80709 and these two uploaded files.

```
[2019-08-06 01:45:27] [Shinobu] [info] All done! Now dutifully watching your files. 
[2019-08-06 01:45:31] [Shinobu] [debug] Received inotify event 256 on /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip
[2019-08-06 01:45:31] [Shinobu] [debug] Adding /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip to Shinobu filemap.
[2019-08-06 01:45:31] [Shinobu] [debug] Computed ID is da39a3ee5e6b4b0d3255bfef95601890afd80709.
[2019-08-06 01:45:31] [Shinobu] [info] Building JSON cache...This might take some time.
[2019-08-06 01:45:32] [Shinobu] [debug] Received inotify event 256 on /db/[Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital].zip
[2019-08-06 01:45:32] [Shinobu] [debug] Adding /db/[Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital].zip to Shinobu filemap.
[2019-08-06 01:45:32] [Shinobu] [debug] Computed ID is da39a3ee5e6b4b0d3255bfef95601890afd80709.
[2019-08-06 01:45:32] [Shinobu] [warn] /db/[Nekodanshaku] Onegai! Onee-chan Please! Onee-chan (COMIC LO 2019-07) [English] [SakuraCircle] [Digital].zip is a duplicate of the existing file /db/(C73) [Sexual Khorosho (Lasto)] Yoshikawa-san to Sugawara-kun! [English] [JnTo].zip. You should delete it.
[2019-08-06 01:45:32] [Shinobu] [info] Building JSON cache...This might take some time.
```
It looks like whenever I upload from web uploader, web uploader properly adds the archive to Redis but Shinobu calculates ID as 'da39a3ee5e6b4b0d3255bfef95601890afd80709' always.

**Comment by CirnoT** [comment]
Looks like IN_CLOSE_WRITE can not be relied upon in many cases. This really is ugly issue, the best solution here would be to ensure that file is saved as .upload and then moved back to original filename all on the same filesystem. Alternatively, something like this should work too:

```
diff --git a/lib/Shinobu.pm b/lib/Shinobu.pm
index 8ddc226..ba73566 100644
--- a/lib/Shinobu.pm
+++ b/lib/Shinobu.pm
@@ -55,7 +55,7 @@ my $inotifysub = sub {
     $logger->debug("Received inotify event $mask on $name");

     if ( $e->IN_MOVED_TO || $e->IN_CREATE ) {
-        new_file_callback( $e->fullname );
+        new_file_callback( $name );
     }

     if ( $e->IN_DELETE ) {
@@ -158,6 +158,14 @@ sub add_to_filemap {
             sleep(1);
         }

+        my $cnt = 0;
+        while (1) {
+            last if (((-s $file) >= 512000) || $cnt >= 60); # wait for file to be more than 512 KBs or bailout after 60s and assume that file is smaller
+            $logger->debug("Waiting for file to be fully written");
+            sleep(1);
+            $cnt++;
+        }
+
         #Compute the ID of the archive and add it to the hash
         my $id = "";
         eval { $id = LANraragi::Utils::Database::compute_id($file); };
@@ -168,6 +176,11 @@ sub add_to_filemap {
             return;
         }

+        if( $id == 'da39a3ee5e6b4b0d3255bfef95601890afd80709' ) {
+            $logger->error("Calculated ID is invalid, giving up.");
+            return;
+        }
+
         $logger->debug("Computed ID is $id.");

         #If the hash already exists, throw a warning about duplicates
@@ -359,3 +372,4 @@ sub build_archive_JSON {
 __PACKAGE__->initialize_from_new_process unless caller;

 1;
```

**Comment by Difegue** [comment]
Ohhhhh now I remember this bug, thought I had solved it with cf20ead5e48a68d98d2cd9917ffee18f9e640333 but it is obviously not enough.  

.tmp files are probably the best bet, since inotify only procs on select extensions the file can be uploaded to the content folder and renamed right at the end.  

I'll handle this post-vacation.
